### PR TITLE
Add parsing metadata from the string

### PR DESCRIPTION
**Pull request by alexsvetlichnyy**

### CODE REVIEW

File: include/COLA.hh [hunk 1]
```diff
@@ -185,7 +193,8 @@ namespace cola {
         ~MetaProcessor() = default;
 
         void reg(std::shared_ptr<VFactory> factory, const std::string& name, const std::string& type);
-        FilterAnsamble parse(const MetaData& data);
+        FilterAnsamble parse(const SimpleMetaData& data);
```
Comment (apBUSampK) [inline include/COLA.hh]: I don't see any usecase for SimpleMetaData struct. Maybe remove it entirely?

File: include/COLA.hh [hunk 2]
```diff
@@ -195,36 +204,68 @@ namespace cola {
...
+    inline cola::MetaData cola::MetaProcessor::parseStrToMeta(std::string data) {
+        std::vector<std::string> filters; cola::MetaData metaData;
+        boost::split(filters, data, boost::is_any_of("\n"));
+        for(int flt = 0; flt < filters.size(); ++flt){
+            std::vector<std::string> tmp;
+            boost::split(tmp, filters.at(flt), boost::is_any_of(" "));
+            std::string res = "";
+            for(int i = 1; i < tmp.size(); ++i){res+=tmp.at(1);}
+            metaData.filterParamMap.emplace(tmp.at(0),res);
+            if(flt == 0){metaData.generatorName = tmp.at(0);}
```
Comment (apBUSampK) [inline include/COLA.hh]: please, remove redundant curved braces for one-line if-else/for statements and add spacing. Or even better, make it indented like this:
```
if (0 < 1)
    std::cout << 1;
```

File: include/COLA.hh [hunk 3]
```diff
@@ -195,36 +204,68 @@ namespace cola {
...
     }
 
+    inline cola::MetaData cola::MetaProcessor::parseStrToMeta(std::string data) {
+        std::vector<std::string> filters; cola::MetaData metaData;
+        boost::split(filters, data, boost::is_any_of("\n"));
+        for(int flt = 0; flt < filters.size(); ++flt){
+            std::vector<std::string> tmp;
+            boost::split(tmp, filters.at(flt), boost::is_any_of(" "));
+            std::string res = "";
+            for(int i = 1; i < tmp.size(); ++i){res+=tmp.at(1);}
```
Comment (apBUSampK) [inline include/COLA.hh]: consider using iterators (std::vector<T>::iterator) or range-based loops. Thery are a part of c++11 standart and are bset practices when working with STL containers

File: include/COLA.hh [hunk 4]
```diff
@@ -195,36 +204,68 @@ namespace cola {
...
+        return ansamble;
+    }
+
+    inline cola::FilterAnsamble cola::MetaProcessor::parse(std::string data) {
         FilterAnsamble ansamble;
-        ansamble.generator = std::shared_ptr<VGenerator>(dynamic_cast<VGenerator*>(generatorMap.at(data.generatorName)->create()));
-        ansamble.converters.push(std::shared_ptr<VConverter>(dynamic_cast<VConverter*>(converterMap.at(data.converterName)->create())));
-        ansamble.writer = std::shared_ptr<VWriter>(dynamic_cast<VWriter*>(writerMap.at(data.writerName)->create()));
+        MetaData meta = parseStrToMeta(data);
+        ansamble.generator = std::shared_ptr<VGenerator>(dynamic_cast<VGenerator*>(generatorMap.at(meta.generatorName)->create(meta.filterParamMap.at(meta.generatorName))));
```
Comment (apBUSampK) [inline include/COLA.hh:207]: I think it is a good idea to move this part to `parse (const cola::MetaData&)` method and just call it here. This way exposing MetaData class to user actually serves some purpose and `parse(std::string)` becomes nice syntactic sugar.

Comment (apBUSampK) [reply include/COLA.hh:207]: Also, maybe change signature to `parse(const std::string)`?

File: include/COLA.hh (hunk 3 above)
Comment (apBUSampK) [reply include/COLA.hh]: It also allows to avoid blunders like .at(1) instead of at(i) ;)

**Review by apBUSampK — CHANGES_REQUESTED** [review]

File: include/COLA.hh (hunk 3 above)
Comment (apBUSampK) [reply include/COLA.hh]: I also actually don't get the purpose of this. Why should we remove all spaces from a substring for a filter? How should a Factory parse it? Isn't it actually up to VFactory implementation to decide what to do with a substring?

File: include/COLA.hh (hunk 1 above)
Comment (alexsvetlichnyy) [reply include/COLA.hh]: I will remove it just before the approval

File: include/COLA.hh (hunk 3 above)
Comment (alexsvetlichnyy) [reply include/COLA.hh]: It is a bug.

File: include/COLA.hh (hunk 4 above)
Comment (alexsvetlichnyy) [reply include/COLA.hh:207]: you want to have it immutable by design?

File: include/COLA.hh (hunk 2 above)
Comment (alexsvetlichnyy) [reply include/COLA.hh]: It is a part of the if else statement

File: include/COLA.hh (hunk 3 above)
Comment (alexsvetlichnyy) [reply include/COLA.hh]: fixed

File: include/COLA.hh (hunk 4 above)
Comment (apBUSampK) [reply include/COLA.hh:207]: Yes, why change the string being parsed?

File: include/COLA.hh (hunk 2 above)
Comment (apBUSampK) [reply include/COLA.hh]: I consulted Google C++ Style and it insists on using spaces and not using compact version of if-else statements.
image

File: include/COLA.hh (hunk 4 above)
Comment (alexsvetlichnyy) [reply include/COLA.hh:207]: done

File: include/COLA.hh (hunk 2 above)
Comment (alexsvetlichnyy) [reply include/COLA.hh]: done

**Pull request merged**